datapath: Improved BPF testing framework

dylandreimerink opened the pull request:
This PR adds a new BPF testing framework for datapath code. This
framework will allow us to write tests in C which will be executed,
as eBPF programs in the kernel. This allows us to test our code against
the actual kernel facilities and helper calls as well as confirming that
code will pass the verifier.

This framework differs from other related works in the sense that we do
the test setup, execution and verification all in C/eBPF code instead
of having to do half the work in userspace and the other half in eBPF.
A lightweight loader program is used to automatically detect test
programs in ELF files in a given directory, load, and execute them using
the `BPF_PROG_TEST_RUN` feature. The test results are passed back to the
loader which will convert them into sub-tests within the golang testing
framework to allow for easy integration into existing tooling.

Doing setup in C allows us to easily mock out code with `#define`
preprocessor tricks, replace tailcalls with stubs/code, and leverage
conditional testing depending on which flags are set / features enabled.
It also decouples datapath testing from the agent allowing us to verify
that the eBPF programs work as expected separately from the agent /
userspace.

Each ELF file can contain multiple `CHECK`'s each of which will become
a separate program which can fail(`test_fail`/`test_fail_now()`) or pass
(default if not failed). Tests can log messages with `test_log("msg")`
for debugging purposes or as fail message(`test_fatal("msg")` logs and
`test_fail_now()`). These messages can include parameters analogous to
`bpf_trace_printk`, for example: 
`test_log("Expected 123, found: %llu", some_var)`. Or for compact 
checks the `assert(some_var == 123)` style can be used which will 
report the file and line of failed asserts and stop the test.

These check programs can also define sub-tests with a `TEST("name", {`
block which can pass/fail independently from the parent test. The appeal
is that such sub-tests run after each other in the exact same context
and can for example test different aspects of the same test setup(ctx,
map state).

Both `CHECK` and `TEST` can be skipped by calling `test_skip()`/
`test_skip_now()`, to mark a test as skipped, for example if applies to
a feature which is disabled or depends on features which are not
available in the current kernel version. Skipped tests are explicitly
marked as such instead of silently ignored. Tests can use `#ifdef`
preprocessor statements to determine to skip or not.

Tests that do not require tailcalls can do setup, execution and
pass/fail checking all in the `CHECK` program. Tests that do need to
tests across multiple tailcalls need to use a `SETUP` program in
addition, in which case the setup and execution will happen in the
`SETUP` program which will exit after the last tailcall is done. The
loader will run any `SETUP` program before the `CHECK` program of the
same name, the `CHECK` program will be invoked with the result context
of the `SETUP` program, the result number of the `SETUP` program will
be pretended to the context(4 bytes), this allows the `CHECK` program
to verify the return value, context and map state and determine a pass/
fail result.

Fixes: #18480

dylandreimerink committed:
bpf: Add `BPF_PROG_TEST_RUN` based testing framework

This commit adds a new BPF testing framework for datapath code. This
framework will allow us to write unit tests in C which will be executed,
as eBPF programs in the kernel. This allows us to test our code against
the actual kernel facilities and helper calls as well as confirming that
code will pass the verifier.

This framework differs from other related works in the sense that we do
the test setup, execution and verification all in C/eBPF code instead
of having to do half the work in userspace and the other half in eBPF.
A lightweight loader program is used to automatically detect test
programs in ELF files in a given directory, load, and execute them using
the `BPF_PROG_TEST_RUN` feature. The test results are passed back to the
loader which will convert them into sub-tests within the golang testing
framework to allow for easy integration into existing tooling.

Doing setup in C allows us to easily mock out code with `#define`
preprocessor tricks, replace tailcalls with stubs/code, and leverage
conditional testing depending on which flags are set / features enabled.
It also decouples datapath testing from the agent allowing us to verify
that the eBPF programs work as expected separately from the agent /
userspace.

Each ELF file can contain multiple `CHECK`'s each of which will become
a separate program which can fail(`test_fail`/`test_fail_now()`) or pass
(default if not failed). Tests can log messages with `test_log("msg")`
for debugging purposes or as fail message(`test_fatal("msg")` logs and
`test_fail_now()`). These messages can include parameters analogous to
`bpf_trace_printk`, for example: `test_log("Expected 123, found: %llu",
some_var)`. Or for compact checks the `assert(some_var == 123)` style
can be used which will report the file and line of failed asserts and
stop the test.

These check programs can also define sub-tests with a `TEST("name", {`
block which can pass/fail independently from the parent test. The appeal
is that such sub-tests run after each other in the exact same context
and can for example test different aspects of the same test setup(ctx,
map state).

Both `CHECK` and `TEST` can be skipped by calling `test_skip()`/
`test_skip_now()`, to mark a test as skipped, for example if applies to
a feature which is disabled or depends on features which are not
available in the current kernel version. Skipped tests are explicitly
marked as such instead of silently ignored. Tests can use `#ifdef`
preprocessor statements to determine to skip or not.

Tests that do not require tailcalls can do setup, execution and
pass/fail checking all in the `CHECK` program. Tests that do need to
tests across multiple tailcalls need to use a `SETUP` program in
addition, in which case the setup and execution will happen in the
`SETUP` program which will exit after the last tailcall is done. The
loader will run any `SETUP` program before the `CHECK` program of the
same name, the `CHECK` program will be invoked with the result context
of the `SETUP` program, the result number of the `SETUP` program will
be pretended to the context(4 bytes), this allows the `CHECK` program
to verify the return value, context and map state and determine a pass/
fail result.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

joamaki commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,443 @@
+
+	for i := 0; i < end; {
+		lasti := i
+		for i < end && l.Fmt[i] != '%' {
```
sometimes code like this reads better when it's "reducing" a slice instead of looping with an index, e.g.
```
fmt := string(l.Fmt)
for len(fmt) > 0 {
  procIndex := strings.Index(fmt, "%")
  if procIndex >= 0 {
    sb.WriteString(fmt[:procIndex])
    fmt = fmt[:procIndex]  
  } else {
    sb.WriteString(fmt)
    break
  }
  ...
}
```

Not suggesting you should do it here, but perhaps food for thought.

nbusseneau commented on .github/workflows/lint-bpf-checks.yaml:
```diff
@@ -115,7 +115,7 @@ jobs:
       - name: Run BPF_PROG_TEST_RUN tests
         run: |
-          make -C bpf go_prog_test || (echo "Run 'make -C bpf go_prog_test' locally to investigate failures"; exit 1)
+          make -C test run_bpf_tests || (echo "Run 'make -C test run_bpf_tests' locally to investigate failures"; exit 1)
```
`lint-bpf-checks.yaml` runs on `pull_request: {}` so no need to do anything :)

joestringer commented on bpf/tests/bpf_tests/bpf_ct_tests.c:
```diff
@@ -0,0 +1,239 @@
+#include "common.h"
```
Ah right, it's just in the other commit. Can you fold these into the original commit?

joestringer commented on test/Makefile:
```diff
@@ -65,3 +65,14 @@ clean:
 	-$(QUIET) rm -f .vagrant/*.box
+
+compile_bpf_tests:
+	$(QUIET) make -C ../bpf/tests/bpf_tests clean
```
Does this need to clean every time? If the targets are set up with dependencies correctly, then these should compile iff something actually changed.

joestringer commented on test/Makefile:
```diff
@@ -65,3 +65,14 @@ clean:
+run_bpf_tests: compile_bpf_tests
+	sudo /tmp/bpf-tests -bpf-test-path ../bpf/tests/bpf_tests
+
+run_bpf_tests_verbose: compile_bpf_tests
```
These new targets that aren't files should be marked `.PHONY`.

joestringer commented on test/Makefile:
```diff
@@ -65,3 +65,14 @@ clean:
+	sudo /tmp/bpf-tests -bpf-test-path ../bpf/tests/bpf_tests
+
+run_bpf_tests_verbose: compile_bpf_tests
+	sudo /tmp/bpf-tests -bpf-test-path ../bpf/tests/bpf_tests -test.v
```
FWIW we have a `V` variable in the Makefiles so you can do something like

```
BPF_TEST_FLAGS:=
ifneq ($(V),0)
    BPF_TEST_FLAGS += -test.v
endif
```

So then if someone explicitly requests no verbosity (`make -C test run_bpf_tests V=0`), it will be quiet. Otherwise it's loud.

This is an option, don't have to do it this way though.

joestringer commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,443 @@
+			}
+
+			if globalLogReader != nil {
+				// Give the global log buf some time to empty
+				time.Sleep(50 * time.Millisecond)
```
:thinking:

dylandreimerink replied:
cilium/ebpf currently doesn't provide this, though I agree it should, and I believe it is possible, I will see if I can fix that. Would it be acceptable to leave it like this and fix it as soon as a flush feature has been upstreamed in cilium/ebpf? (if that takes longer than this PR)

joestringer replied:
Maybe a question first, what's the consequences if 50 Milliseconds is not enough time to flush the log? Can we easily detect that problem and react to it?

joestringer commented on test/bpf_tests/trf.proto:
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: Apache-2.0
```
I'm curious about interactions using this trf format from the test prog to other systems. Is this just a nice way to get output formatted in a certain way, or can these be exported into some well-known system? How will we use it? This wasn't obvious to me from the commit message.

dylandreimerink replied:
I initially wrote my own serialization format, with the serializer in eBPF and a janky deserializer in Go. But I felt that creating a custom serialization format for a test suite was overkill and bloat. 

So I decided to use protobuf, which is compact on the wire due to the external `.proto` definitions which actually make it easier to craft messages in eBPF than something like JSON or MessagePack. And the Go code is free with ProtoBuf since all structs and deserialization is generated. Plus we already use protobuf in other places in the codebase so it doesn't add any dependencies. Seemed like a good choice all around.

joestringer replied:
That seems like a good fit. How will we use this output? Do developers just read the protobuf output in a file somewhere?

joestringer replied:
:100: nice.

joestringer reviewed (commented):
It looked like others focused on the details of the test progs themselves, so I focused more on the build systems and general bits that were different from the way that other go tests work in the tree.

I explicitly didn't review the actual test content, because if you delete the full files and then add full files, it's much more difficult to review the diffs. If you're able to combine these in the same commit with minimal diff, that would make review easier for those pieces.

dylandreimerink commented on bpf/tests/bpf_tests/common.h:
```diff
@@ -0,0 +1,239 @@
+// https://developers.google.com/protocol-buffers/docs/encoding#structure
+
+// Field 1, Type 2
+#define MKR_TEST_RESULT 0b0001010
```
Yea, a macro seems like a way better idea, should have thought of that. 

> an ebpf protoc plugin would be great

Agreed, I already have to much side projects though, perhaps one day

dylandreimerink commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,443 @@
+
+			coll, err := ebpf.NewCollectionWithOptions(spec, ebpf.CollectionOptions{
+				Programs: ebpf.ProgramOptions{
+					LogSize: 1 << 20,
```
Agreed on both counts, creating a loop to grow the log size if ENOSPC is encountered is trivial, have done this before. Then again, do we want more than 1MB of verifier log in the CI output? Perhaps we can do something here with flags for size and verbosity so we don't get that much output in the CI, but we can get detailed info when running locally to fix a detected issue.

As for how to handle this in general... It depends. It would be nice if we could incorporate this behavior in cilium/ebpf by specifying size -1 or something like it. But it is not something you want all the time, because of memory consumption and such.

dylandreimerink commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,443 @@
+				linkCache := link.NewLinkCache()
+
+				// Closes as soon as the goroutine has started
+				started := make(chan struct{})
```
No, I don't think so. Since any messages accumulated should be buffered in the perf buffer after we called `perf.NewReader`. I think I was confused because `globalLogReader.Read` is blocking so I assumed the messages would get lost if there was not blocking goroutine ready. Will fix that.

dylandreimerink commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,443 @@
+			switch c {
+			case 'd', 'i', 'u', 'x', 's':
+				spec = append(spec, c)
+				goto specloopend
```
I didn't know this was an option, cool.

dylandreimerink commented on bpf/tests/bpf_tests/Makefile:
```diff
@@ -0,0 +1,49 @@
+# Copyright Authors of Cilium
```
Yes, you are right. Will fix that. I did it like this because when I started we had other tests in the `bpf/tests` dir with would mess with what I was doing. Not needed anymore since I have replaced all existing tests.

dylandreimerink commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,443 @@
+// SPDX-License-Identifier: Apache-2.0
```
Currently the distinction seems to be that `/bpf` is for C code and the rest for Go. You can't put Go and C files in the same directory since Go will automatically attempt to compile .c files as CGO, so any Go code would have to live in a separate directory anyway. Having 1 Go file in among all the C code felt strange, but I agree that it would be better to co-locate code that works together, perhaps not just for the tests but other parts of cilium as well

dylandreimerink committed:
bpf: Allow uneven numbers in `memcpy` and `memcmp`

The `memcpy` and `memcmp` builtin replacements did work for uneven
values other than 1. This commit updates both builtin replacements so
they do. This is quite important since it isn't uncommon to copy strings
with an uneven amount of chars in the new test framework.

Additionally, this commit extends the max size of values passed into
`memcmp` from 32 bytes to 72 bytes which is required for some unit tests
to compare packets which are larger than 32 bytes.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

brb reviewed (approved):
Checked the test conversion + new tests, LGTM!

ti-mo commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,470 @@
+	}
+}
+
+func loadAndPrepSpec(t *testing.T, elfPath string) *ebpf.CollectionSpec {
```
(see my comment below on `cilium_xdp_scratch`. What about pulling in the first two commits of https://github.com/cilium/cilium/pull/19159 and building on top of those instead? It already includes the logic around prog types (although with a different heuristic) and MapSpec.Extra.

It also solves populating the calls map based on `X/Y` section names.

dylandreimerink replied:
I really wouldn't mind if we didn't complicating the PR any more than we have to. I would prefer to merge this and then iterate on it.

ti-mo replied:
Fine by me, as long as it gets done. :) It would eliminate much of the code in this file.

ti-mo commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,470 @@
+	suiteResultMap = "suite_result_map"
+
+	callsMapName = "test_cilium_calls_65535"
+	// TODO we should read the .id field from the maps BTF, but the current cilium/ebpf version doesn't make the raw
```
The .id field is not emitted into BTF, it's in the first 4 bytes of MapSpec.Extra. Again, https://github.com/cilium/cilium/pull/19159/commits/7aaec16d6076e6f6ac562bfc4bdaf2650a42733d takes care of that.

ti-mo commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,470 @@
+	args []uint64
+}
+
+// A simplified version of fmt.Printf logic, the meaning of % specifiers changed to match the kernels printk specifiers.
```
Could this contain some examples? It's not very clear what this does. Why is it called `FmtString()`?

dylandreimerink replied:
I added an example and references, I hope that makes it clearer. The reason for calling it `FmtString()` is because `String()` is already taken. I define a function on the `Log` struct which is defined by the protobuf generated code, which itself implements `String()` by default, but that of course doesn't do the string formatting magic we want.

ti-mo reviewed (commented):
Did a quick review of the Go bits. Please take from https://github.com/cilium/cilium/pull/19159 what you can (prob first 2 commits) to de-duplicate the effort/code. The reason those haven't been merged separately is because there are no other consumers of that code.

dylandreimerink commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,470 @@
+	var (
+		coll    *ebpf.Collection
+		err     error
+		logSize = 1 << 20 // 1 MiB
```
Perhaps this code is somewhat over-engineered at the moment. I personally think this isn't a bad approach considering that versifier logs will significantly grow once bounded loops, bpf-to-bpf functions and 1M instruction limit become standard in the cilium codebase. Leaving the 1MiB verifier log as a landmine for the poor soul who has to find it at the worst possible moment seems lazy to me.

joestringer commented on bpf/tests/bpf_tests/bpf_ct_tests.c:
```diff
@@ -0,0 +1,239 @@
+#include "common.h"
+
+#include <bpf/ctx/skb.h>
```
Yeah I think for that to work, you need to explicitly use `git mv bpf/foo.h bpf/foo.c` and then make the changes on top.

joestringer commented on bpf/tests/Makefile:
```diff
@@ -0,0 +1,48 @@
+
+FLAGS := -I$(ROOT_DIR)/bpf/include -I$(ROOT_DIR)/bpf -D__NR_CPUS__=$(shell nproc --all) -O2 -g
+
+CLANG_FLAGS := ${FLAGS} -target bpf -std=gnu99 -nostdinc -emit-llvm
+# eBPF verifier enforces unaligned access checks where necessary, so don't
+# let clang complain too early.
+CLANG_FLAGS += -Wall -Wextra -Werror -Wshadow
+CLANG_FLAGS += -Wno-address-of-packed-member
+CLANG_FLAGS += -Wno-unknown-warning-option
+CLANG_FLAGS += -Wno-gnu-variable-sized-type-not-at-end
+CLANG_FLAGS += -Wdeclaration-after-statement
+CLANG_FLAGS += -Wimplicit-int-conversion -Wenum-conversion
+LLC_FLAGS := -march=bpf -mattr=dwarfris
```
Do we need to be figuring out the clang flags independently for the test progs? Or can we reuse existing flags from the other makefiles? (even if we have to shift them to either `Makefile.defs` or `bpf/Makefile.bpf` or something).

ti-mo commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,470 @@
+	// Collect debug events and add them as logs of the main test
+	var globalLogReader *perf.Reader
+	if m := coll.Maps["test_cilium_events"]; m != nil {
+		globalLogReader, err = perf.NewReader(m, 4096*16)
```
```suggestion
		globalLogReader, err = perf.NewReader(m, os.Getpagesize()*16)
```

YutaroHayakawa commented on bpf/tests/xdp_nodeport_lb4_test.c:
```diff
@@ -0,0 +1,214 @@
+		},
+};
+
+#define BACKEND_IP 0x0F00020A // 10.2.0.10
```
Nit: `10.2.0.15`?

YutaroHayakawa replied:
```suggestion
#define BACKEND_IP 0x0F00020A // 10.2.0.15
```

YutaroHayakawa commented on bpf/tests/xdp_nodeport_lb4_test.c:
```diff
@@ -0,0 +1,214 @@
+		.frag_off = 0x4000,
+		.ttl = 64,
+		.protocol = IPPROTO_TCP,
+		.saddr = 0x0F00000A, // 10.0.0.16
+		.daddr = 0x0F00010A // 10.0.1.16
```
Nit: `10.0.0.15` & `10.1.0.15`?

YutaroHayakawa replied:
```suggestion
		.saddr = 0x0F00000A, // 10.0.0.15
		.daddr = 0x0F00010A // 10.1.0.15
```

YutaroHayakawa commented on test/bpf_tests/bpf_test.go:
```diff
@@ -0,0 +1,459 @@
+		t.Fatal("os readdir: ", err)
+	}
+
+	pinPath := "/sys/fs/bpf/cilium-test"
+	defer func() {
+		os.RemoveAll(pinPath)
+	}()
+	if err = os.Mkdir(pinPath, 0755); err != nil {
+		t.Fatal(err)
+	}
```
I'm wondering who should mount this BPFFS? At this moment, there's no logic to mount BPFFS inside `make -C test run_bpf_test`. When I tried it in my local environment with `umount`-ing BPFFS, it produced this error. I think it would be better to mount it in `Makefile` or here if it is not mounted yet. What do you think?

```
$ make -C test run_bpf_tests                                                                                                                                                                                      /home/yutaro/WorkSpace/cilium
make: Entering directory '/home/yutaro/WorkSpace/cilium/test'
make -C ../bpf/tests all
make[1]: Entering directory '/home/yutaro/WorkSpace/cilium/bpf/tests'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/yutaro/WorkSpace/cilium/bpf/tests'
go test bpf_tests/*.go -c -o /tmp/bpf-tests
sudo /tmp/bpf-tests -bpf-test-path ../bpf/tests  -test.v
=== RUN   TestBPF
    bpf_test.go:55: mkdir /sys/fs/bpf/cilium-test: no such file or directory
--- FAIL: TestBPF (0.00s)
FAIL
make: *** [Makefile:79: run_bpf_tests] Error 1
make: Leaving directory '/home/yutaro/WorkSpace/cilium/test'
```

ti-mo replied:
@dylandreimerink Perhaps we should clear all maps' Pinning flags instead? Since this style of test (mainly driven from BPF C) doesn't/shouldn't access bpffs, tests can stay isolated without any risk of leaving leftover state, reducing flake potential.

All maps/progs can be accessed through Collections directly in Go, so I don't think we need to pin anything. bpffs interactions occur purely in userspace through Go code, which kind of falls outside of the framework's scope for now IMO.

dylandreimerink replied:
Yes, I set the `Pinning` setting on all maps to none, so we don't pin the maps during testing.

pchaigno committed:
cocci: Add more special cases for the const script

The const Coccinelle script attempts to detect all function arguments
that should be marked as const. To that end, for each function argument,
we skip all cases that could lead to that argument being modified.

One corner case we need to detect is when we modify an array field in a
struct (`func(..., x->field, ...)`). In that case, we can't easily
distinguish between a pointer field and an array field. In the former
case, the struct is not modified; in the latter case, it is. Since we
only have a few cases with array fields in the codebase, we list those
manually.

Until recently, that was only the addr field in some struct (holding an
IPv6). We now need to add the fields for MAC addresses.

Signed-off-by: Paul Chaignon <paul@cilium.io>

dylandreimerink committed:
bpf: ported existing unit tests to new testing framework

This commit ports most existing unit tests in their different forms to
the new eBPF unit test framework. This removes the need for a mocking
framework in userspace and replaces the existing BPF_PROG_TEST_RUN based
framework in the CI.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
